webapp/platform/client/src/websocket.test.ts

```diff
@@ -0,0 +1,206 @@
+// See LICENSE.txt for license information.
+
+import WebSocketClient from './websocket';
+
+// Define some WebSocket globals that aren't defined in node
+if (typeof WebSocket === 'undefined') {
+    (global as any).WebSocket = {
+        CONNECTING: 0, OPEN: 1, CLOSING: 2, CLOSED: 3,
+    };
+}
+
+export class MockWebSocket {
+    readonly binaryType: BinaryType = 'blob';
+    readonly bufferedAmount: number = 0;
+    readonly extensions: string = '';
+
+    readonly CONNECTING = WebSocket.CONNECTING;
+    readonly OPEN = WebSocket.OPEN;
+    readonly CLOSING = WebSocket.CLOSING;
+    readonly CLOSED = WebSocket.CLOSED;
+
+    public url: string = '';
+    readonly protocol: string = '';
+    public readyState: number = WebSocket.CONNECTING;
+
+    public onopen: (() => void) | null = null;
+    public onclose: (() => void) | null = null;
+    public onerror: (() => void) | null = null;
+    public onmessage: ((evt: any) => void) | null = null;
+
+    open() {
+        this.readyState = WebSocket.OPEN;
+        if (this.onopen) {
+            this.onopen();
+        }
+    }
+
+    close() {
+        this.readyState = WebSocket.CLOSED;
+        if (this.onclose) {
+            this.onclose();
+        }
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    send(msg: any) { }
+    addEventListener() { }
+    removeEventListener() { }
+    dispatchEvent(): boolean {
+        return false;
+    }
+}
+
+describe('websocketclient', () => {
+    test('should call callbacks', () => {
+        const mockWebSocket = new MockWebSocket();
+
+        const client = new WebSocketClient({
+            newWebSocketFn: (url: string) => {
+                mockWebSocket.url = url;
+                return mockWebSocket;
+            },
+        });
+        client.initialize('mock.url');
+
+        expect(mockWebSocket.onopen).toBeTruthy();
+        mockWebSocket.onopen = jest.fn();
+        expect(mockWebSocket.onclose).toBeTruthy();
+        mockWebSocket.onclose = jest.fn();
+
+        mockWebSocket.open();
+
+        expect(mockWebSocket.onopen).toHaveBeenCalled();
+        expect(mockWebSocket.readyState).toBe(mockWebSocket.OPEN);
+
+        mockWebSocket.close();
+
+        expect(mockWebSocket.onclose).toHaveBeenCalled();
+        expect(mockWebSocket.readyState).toBe(mockWebSocket.CLOSED);
+
+        client.close();
+    });
+
+    test('should reconnect on websocket close', (done) => {
+        const mockWebSocket = new MockWebSocket();
+        mockWebSocket.open = jest.fn(mockWebSocket.open);
+
+        const client = new WebSocketClient({
+            newWebSocketFn: (url: string) => {
+                mockWebSocket.url = url;
+                mockWebSocket.open();
+                return mockWebSocket;
+            },
+            minWebSocketRetryTime: 1,
+            reconnectJitterRange: 1,
+        });
+        client.initialize('mock.url');
+
+        mockWebSocket.close();
+
+        setTimeout(() => {
```

Comment:
Instead of making these tests use arbitrary `setTimeout`s, you may be able to write these to use Jest's fake timers (https://jestjs.io/docs/timer-mocks) to avoid the arbitrary delays and reduce the chances that something causes the timeouts to run out of order.

I probably went a bit overboard with them in this file, but it's at least an example of how to use them: https://github.com/mattermost/mattermost/blob/master/webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.test.ts